### PR TITLE
extmod/uasyncio: Fix edge case for cancellation of wait_for.

### PR DESCRIPTION
**Pull request by dpgeorge**
Fixes issue #8717.

### CODE REVIEW

File: extmod/uasyncio/funcs.py [hunk 1]
```diff
@@ -10,40 +10,41 @@ async def wait_for(aw, timeout, sleep=core.sleep):
...
+            status = True
         except BaseException as er:
-            s = er
-        if status is None:
-            # The waiter is still waiting, set status for it and cancel it.
-            status = s
-            waiter.cancel()
+            result = None
+            status = er
+        if waiter.data is None:
```
Comment (Kalki70) [inline extmod/uasyncio/funcs.py:19]: Wouldn't it be better to check that it is not instance of core.CancelledError?

Comment (dpgeorge) [reply extmod/uasyncio/funcs.py:19]: I tried this, some tests will fail.  Eg when `waiter` is waiting on `runner` (in that case `waiter.data` will point to the `runner` task).

**Review by Kalki70 — APPROVED** [review]
The changes look OK, and it is not possible to replicate the faulty scenario  anymore with this fix.

**Comment by dpgeorge** [comment]
> thought the idea of next line was to make sure that a cancelled task would be scheduled for execution immediately after it was cancelled :

This line reschedules the task to run the cancellation ASAP, but it still puts it after any existing tasks that are already scheduled to run next (ie whose time delta is <= 0).  That's necessary to ensure fairness of the scheduler.  Otherwise you could keep cancelling a task and starve other tasks.

> I get a bit lost about the use of Task.data, as it sometimes contains different type of data.

Yes, that's done for (memory) efficiency reasons.  MicroPython needs to run in constrained places without much RAM and so we must sometimes use tricks in the implemenation.  The internals of uasyncio can be a bit tricky, but that's all hidden from the user of the library.

Thanks for reviewing and testing!

**Comment by dpgeorge** [comment]
I improved it by putting the `runner()` function in the global scope, because it's no longer a closure and doesn't need to be created inside the `wait_for` function anymore.

**Comment by Kalki70** [comment]
> Thanks for reviewing and testing!

Thanks for the answers and the quick fix!

**Pull request merged**